client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -0,0 +1,783 @@
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! API implementation for `chainHead`.
+
+use crate::{
+	chain_head::{
+		api::ChainHeadApiServer,
+		error::Error as ChainHeadRpcError,
+		event::{
+			BestBlockChanged, ChainHeadEvent, ChainHeadResult, ErrorEvent, Finalized, FollowEvent,
+			Initialized, NetworkConfig, NewBlock, RuntimeEvent, RuntimeVersionEvent,
+		},
+		subscription::{SubscriptionHandle, SubscriptionManagement, SubscriptionManagementError},
+	},
+	SubscriptionTaskExecutor,
+};
+use codec::Encode;
+use futures::{
+	channel::oneshot,
+	future::FutureExt,
+	stream::{self, Stream, StreamExt},
+};
+use futures_util::future::Either;
+use jsonrpsee::{
+	core::{async_trait, RpcResult},
+	types::{SubscriptionEmptyError, SubscriptionResult},
+	SubscriptionSink,
+};
+use log::{debug, error};
+use sc_client_api::{
+	Backend, BlockBackend, BlockImportNotification, BlockchainEvents, CallExecutor, ChildInfo,
+	ExecutorProvider, FinalityNotification, StorageKey, StorageProvider,
+};
+use serde::Serialize;
+use sp_api::CallApiAt;
+use sp_blockchain::{
+	Backend as BlockChainBackend, Error as BlockChainError, HeaderBackend, HeaderMetadata,
+};
+use sp_core::{hexdisplay::HexDisplay, storage::well_known_keys, Bytes};
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use std::{marker::PhantomData, sync::Arc};
+
+/// An API for chain head RPC calls.
+pub struct ChainHead<BE, Block: BlockT, Client> {
+	/// Substrate client.
+	client: Arc<Client>,
+	/// Backend of the chain.
+	backend: Arc<BE>,
+	/// Executor to spawn subscriptions.
+	executor: SubscriptionTaskExecutor,
+	/// Keep track of the pinned blocks for each subscription.
+	subscriptions: Arc<SubscriptionManagement<Block>>,
+	/// The hexadecimal encoded hash of the genesis block.
+	genesis_hash: String,
+	/// The maximum number of pinned blocks allowed per connection.
+	max_pinned_blocks: usize,
+	/// Phantom member to pin the block type.
+	_phantom: PhantomData<Block>,
+}
+
+impl<BE, Block: BlockT, Client> ChainHead<BE, Block, Client> {
+	/// Create a new [`ChainHead`].
+	pub fn new<GenesisHash: AsRef<[u8]>>(
+		client: Arc<Client>,
+		backend: Arc<BE>,
+		executor: SubscriptionTaskExecutor,
+		genesis_hash: GenesisHash,
+		max_pinned_blocks: usize,
+	) -> Self {
+		let genesis_hash = format!("0x{}", hex::encode(genesis_hash));
+
+		Self {
+			client,
+			backend,
+			executor,
+			subscriptions: Arc::new(SubscriptionManagement::new()),
+			genesis_hash,
+			max_pinned_blocks,
+			_phantom: PhantomData,
+		}
+	}
+
+	/// Accept the subscription and return the subscription ID on success.
+	///
+	/// Also keep track of the subscription ID internally.
+	fn accept_subscription(
+		&self,
+		sink: &mut SubscriptionSink,
+	) -> Result<String, SubscriptionEmptyError> {
+		// The subscription must be accepted before it can provide a valid subscription ID.
+		sink.accept()?;
+
+		// TODO: Jsonrpsee needs release + merge in substrate
```

Comment:
Note: This bit will be uncommented as soon as https://github.com/paritytech/substrate/pull/12663 gets merged

Reply:
Yep, the subscriptions need the `subscription_id` feature https://github.com/paritytech/jsonrpsee/pull/900 from the latest release.